### ISSUE TITLE
Does this apply to the llama2 model?

### ISSUE THREAD

**Issue by Tincsvsv** [lancopku/label-words-are-anchors#4]
First of all, congratulations to the authors for winning the best paper award.
Does this apply to the llama2 model? If so, what should be changed?

**Comment by leanwang326** [comment]
To adapt our code to a new model like llama, you should adapt the GPT2AttentionerManager and gpt2_attn (actually, most of our codes is compatible with other methods, but there is one thing we find hard to write a direct code to add masks to `attn_weights` or get gradients of the `attn_weights`). 

So, you can rewrite this part

```
attn_weights = nn.Softmax(dim=-1)(attn_weights)

if attention_adapter is not None:
    attn_weights = attention_adapter(attn_weights)

attn_weights = attn_weights.type(value.dtype)
attn_weights = self.attn_dropout(attn_weights)
```

to get `llama_attn`:

```
def llama_attn(
        self,
        hidden_states: torch.Tensor,
        attention_mask: Optional[torch.Tensor] = None,
        position_ids: Optional[torch.LongTensor] = None,
        past_key_value: Optional[Tuple[torch.Tensor]] = None,
        output_attentions: bool = False,
        use_cache: bool = False,
        attention_adapter=None,
) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[Tuple[torch.Tensor]]]:
    bsz, q_len, _ = hidden_states.size()

    query_states = self.q_proj(hidden_states).view(bsz, q_len, self.num_heads,
                                                   self.head_dim).transpose(1, 2)
    key_states = self.k_proj(hidden_states).view(bsz, q_len, self.num_heads,
                                                 self.head_dim).transpose(1, 2)
    value_states = self.v_proj(hidden_states).view(bsz, q_len, self.num_heads,
                                                   self.head_dim).transpose(1, 2)

    kv_seq_len = key_states.shape[-2]
    if past_key_value is not None:
        kv_seq_len += past_key_value[0].shape[-2]
    cos, sin = self.rotary_emb(value_states, seq_len=kv_seq_len)
    query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin,
                                                    position_ids)
    # [bsz, nh, t, hd]

    if past_key_value is not None:
        # reuse k, v, self_attention
        key_states = torch.cat([past_key_value[0], key_states], dim=2)
        value_states = torch.cat([past_key_value[1], value_states], dim=2)

    past_key_value = (key_states, value_states) if use_cache else None

    attn_weights = torch.matmul(query_states, key_states.transpose(2, 3)) / math.sqrt(self.head_dim)

    if attn_weights.size() != (bsz, self.num_heads, q_len, kv_seq_len):
        raise ValueError(
            f"Attention weights should be of size {(bsz * self.num_heads, q_len, kv_seq_len)}, but is"
            f" {attn_weights.size()}"
        )

    if attention_mask is not None:
        if attention_mask.size() != (bsz, 1, q_len, kv_seq_len):
            raise ValueError(
                f"Attention mask should be of size {(bsz, 1, q_len, kv_seq_len)}, but is {attention_mask.size()}"
            )
        attn_weights = attn_weights + attention_mask
        attn_weights = torch.max(attn_weights, torch.tensor(torch.finfo(attn_weights.dtype).min))

    # print(attn_weights.shape)


    if attention_adapter is not None:
        attn_weights = attention_adapter(attn_weights)

    # upcast attention to fp32
    attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(
        query_states.dtype)
    attn_output = torch.matmul(attn_weights, value_states)

    if attn_output.size() != (bsz, self.num_heads, q_len, self.head_dim):
        raise ValueError(
            f"`attn_output` should be of size {(bsz, self.num_heads, q_len, self.head_dim)}, but is"
            f" {attn_output.size()}"
        )


    attn_output = attn_output.transpose(1, 2)
    attn_output = attn_output.reshape(bsz, q_len, self.hidden_size)

    attn_output = self.o_proj(attn_output)

    if not output_attentions:
        attn_weights = None

    return attn_output, attn_weights, past_key_value
```

And you may adjust the code to use the corresponding LlamaAttentionerManager and llama_attn (and you may remove some assert sentence that assert model_name to be 'gpt2' or 'gptj')



If you face some further questions or have some interesting findings, feel free to contact us!

**Comment by Tincsvsv** [comment]
Thank you for your reply, which is of great help to me!
